cpp/include/binaryop.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BINARYOP_HPP
+#define BINARYOP_HPP
+
+#include <cudf.h>
+#include <types.hpp>
+
+/**
+ * @brief Types of binary operations that can be performed on data.
+ */
+typedef enum {
+  GDF_ADD,            ///< operator +
+  GDF_SUB,            ///< operator -
+  GDF_MUL,            ///< operator *
+  GDF_DIV,            ///< operator / using common type of lhs and rhs
+  GDF_TRUE_DIV,       ///< operator / after promoting type to floating point
+  GDF_FLOOR_DIV,      ///< operator / after promoting to float and then flooring the result
+  GDF_MOD,            ///< operator %
+  GDF_POW,            ///< lhs ^ rhs
+  GDF_EQUAL,          ///< operator ==
+  GDF_NOT_EQUAL,      ///< operator !=
+  GDF_LESS,           ///< operator <
+  GDF_GREATER,        ///< operator >
+  GDF_LESS_EQUAL,     ///< operator <=
+  GDF_GREATER_EQUAL,  ///< operator >=
+  GDF_BITWISE_AND,    ///< operator &
+  GDF_BITWISE_OR,     ///< operator |
+  GDF_BITWISE_XOR,    ///< operator ^
+  GDF_LOGICAL_AND,    ///< operator &&
+  GDF_LOGICAL_OR,     ///< operator ||
+  GDF_COALESCE,       ///< operator x,y  x is null ? y : x
+  GDF_INVALID_BINARY  ///< invalid operation
+} gdf_binary_operator;
+
+namespace cudf
+{
+
+/**
+ * @brief Performs a binary operation between a gdf_scalar and a gdf_column.
+ *
+ * The desired output type must be specified in out->dtype.
+ *
+ * If the valid field in the gdf_column output is not nullptr, then it will be
+ * filled with the bitwise AND of the valid mask of rhs gdf_column and is_valid
+ * bool of lhs gdf_scalar
+ *
+ * @param out (gdf_column) Output of the operation.
+ * @param lhs (gdf_scalar) First operand of the operation.
+ * @param rhs (gdf_column) Second operand of the operation.
+ * @param ope (enum) The binary operator to use
+ */
+void binary_operation_s_v(gdf_column*           out,
```

Comment:
Since you are updating these functions to C++ APIs (vs. the old C-style API), you should:
 - Allocate and return the output column rather than accepting a pre-allocated one.
 - Make the input column and scalar parameters `gdf_column const &` and `gdf_scalar const &`, respectively.

See https://github.com/rapidsai/cudf/blob/branch-0.8/cpp/docs/TRANSITIONGUIDE.md#inputoutput-style

Reply:
Likewise, no need for the `_s_v`, `_v_v` suffixes. This is C++, let's just use normal function overloading.

Reply:
Currently we use the passed output `gdf_column` to know what output dtype to use and we might have to ask for `dtype_info` in the future (for timestamp maybe). Also, adding `const` is going to edit a lot of files that I'm working on in #1245. I'll put off this part of the API refactor to after #1245 is merged and after we have a `cudf::column` class.

Removing suffixes is easy though.